turbo/jsonrpc/trace_filtering.go

```diff
@@ -78,16 +78,21 @@ func (api *TraceAPIImpl) Transaction(ctx context.Context, txHash common.Hash, ga
 	}
 
 	var txIndex int
-	for idx := 0; idx < block.Transactions().Len() && !isBorStateSyncTxn; idx++ {
-		txn := block.Transactions()[idx]
-		if txn.Hash() == txHash {
-			txIndex = idx
-			break
-		}
-	}
-
 	if isBorStateSyncTxn {
 		txIndex = block.Transactions().Len()
+	} else {
+		var found bool
+		for idx := 0; idx < block.Transactions().Len(); idx++ {
+			txn := block.Transactions()[idx]
+			if txn.Hash() == txHash {
+				txIndex = idx
+				found = true
+				break
+			}
+		}
+		if !found {
+			return nil, fmt.Errorf("block with hash %x is not currently canonical. blockNum=%d, txnHash=%x", block.Hash(), block.NumberU64(), txHash)
```

Comment:
I think it's the other way round. Block is canonical, transaction is not. Right?

Reply:
you right